book-content/chapters/16-the-utils-folder.md

```diff
@@ -0,0 +1,1527 @@
+      } & {
+        code: number;
+      }
+    >
+  >;
+
+  type test4 = Expect<Equal<typeof customErrorWithCode.message, string>>;
+});
+```
+
+Your task is to update the `addCodeToError` type signature to enforce the required constraints so that `TError` is required to have a `message` property and can optionally have a `code` property.
+
+<Exercise title="Exercise 4: Type Parameter Constraints" filePath="/src/085-the-utils-folder/216-type-parameter-defaults-in-generic-functions.problem.ts"></Exercise>
+
+### Exercise 5: Combining Generic Types and Functions
+
+Here we have `safeFunction`, which accepts a function `func` typed as `PromiseFunc` that returns a function itself. However, if `func` encounters an error, it is caught and returned instead:
+
+```typescript
+type PromiseFunc = () => Promise<any>;
+
+const safeFunction = (func: PromiseFunc) => async () => {
+  try {
+    const result = await func();
+    return result;
+  } catch (e) {
+    if (e instanceof Error) {
+      return e;
+    }
+    throw e;
+  }
+};
+```
+
+In short, the thing that we get back from `safeFunction` should either be the thing that's returned from `func` or an `Error`.
+
+However, there are some issues with the current type definitions.
+
+The `PromiseFunc` type is currently set to always return `Promise<any>`. This means that the function returned by `safeFunction` is supposed to return either the result of `func` or an `Error`, but at the moment, it's just returning `Promise<any>`.
+
+There are several tests that are failing due to these issues:
+
+```ts twoslash
+// @errors: 2344
+import { Equal, Expect } from "@total-typescript/helpers";
+import { it, expect } from "vitest";
+
+type PromiseFunc = () => Promise<any>;
+
+const safeFunction = (func: PromiseFunc) => async () => {
+  try {
+    const result = await func();
+    return result;
+  } catch (e) {
+    if (e instanceof Error) {
+      return e;
+    }
+    throw e;
+  }
+};
+
+// ---cut---
+it("should return an error if the function throws", async () => {
+  const func = safeFunction(async () => {
+    if (Math.random() > 0.5) {
+      throw new Error("Something went wrong");
+    }
+    return 123;
+  });
+
+  type test1 = Expect<Equal<typeof func, () => Promise<Error | number>>>;
+
+  const result = await func();
+
+  type test2 = Expect<Equal<typeof result, Error | number>>;
+});
+
+it("should return the result if the function succeeds", async () => {
+  const func = safeFunction(() => {
+    return Promise.resolve(`Hello!`);
+  });
+
+  type test1 = Expect<Equal<typeof func, () => Promise<string | Error>>>;
+
+  const result = await func();
+
+  type test2 = Expect<Equal<typeof result, string | Error>>;
+
+  expect(result).toEqual("Hello!");
+});
+```
+
+Your task is to update `safeFunction` to have a generic type parameter, and update `PromiseFunc` to not return `Promise<Any>`. This will require you to combine generic types and functions to ensure that the tests pass successfully.
```

Comment:
```suggestion
Your task is to update `safeFunction` to have a generic type parameter, and update `PromiseFunc` to not return `Promise<any>`. This will require you to combine generic types and functions to ensure that the tests pass successfully.
```